OpenSim/Common/TableSource.h

```diff
@@ -0,0 +1,214 @@
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#ifndef OPENSIM_TABLE_SOURCE_H_
+#define OPENSIM_TABLE_SOURCE_H_
+
+#include "TimeSeriesTable.h"
+#include "Component.h"
+
+namespace OpenSim {
+
+class TimeOutOfRange : public InvalidTimestamp {
+public:
+    TimeOutOfRange(const std::string& file,
+                   size_t line,
+                   const std::string& func,
+                   double timestamp,
+                   double minTimestamp,
+                   double maxTimestamp) :
+        InvalidTimestamp(file, line, func) {
+        std::string msg = "min = " + std::to_string(minTimestamp);
+        msg += " max = " + std::to_string(maxTimestamp);
+        msg += " timestamp = " + std::to_string(timestamp);
+
+        addMessage(msg);
+    }
+}; // class TimeOutOfRange
+
+
+/** Component representing a source of data from a TimeSeriesTable_.
+
+This Component has two outputs:
+- A list output wih one channel per column of the TimeSeriesTable_. 
+- A non-list output for a row of the TimeSeriesTable_.
+
+Construct this Component by giving it a TimeSeriesTable_. Then use it by 
+connecting its output to the Input of another Component that accepts compatible
+type of Input. Make sure to populate the column-labels of the TimeSeriesTable_ 
+before connecting this Component to the input of another Component.    
+
+\tparam ET Type of each element of the TimeSeriesTable_ this Component 
+           represents.                                                        */
+template<typename ET>
+class TableSource_ : public Component {
+    OpenSim_DECLARE_CONCRETE_OBJECT_T(TableSource_, ET, Component);
+
+public:
+    /** Type of the TimeSeriesTable_ this Component will hold.                */
+    typedef TimeSeriesTable_<ET> Table;
+    /** Type of the 'row' Output of this Component.                           */
+    typedef SimTK::Vector_<ET>   Vector;
+
+    OpenSim_DECLARE_OUTPUT(row, Vector, getRowAtTime, 
+                           SimTK::Stage::Time);
+    OpenSim_DECLARE_LIST_OUTPUT(column, ET, getColumnAtTime, 
+                                SimTK::Stage::Time);
```

Comment:
I agree. The list output also requires the index or name at access time, whereas the Vector of all values does not, but they both potentially can provide all the output values to satisfy a corresponding list input (e.g. [test-case](https://github.com/opensim-org/opensim-core/pull/935/files#diff-d4268f0f931b28653836e567b134a6b8R170) ) in which case a singular "column" seems incorrect. How about `vector` (for the vector of values)  vs `elements` (for many values accessible as individual elements)? I don't have my heart set on these. 

@tkuchida @jimmyDunne  @jenhicks any suggestions here?